File: May-LeetCoding-Challange-2020/Week1/169-majority-element.cpp
class Solution {
public:
    
    /* Time Complexity: O(n)
     * Space Complexity: O(n)
    */
    int majorityElement(vector<int>& nums) {
        int n = nums.size();
        
        if(n == 0 || n == 1){
            return nums[0];
        }
        
        map<int, int> a;
        int maj;
        
        for(int i = 0; i < n; i++){
            a[nums[i]]++;
            
            if(a[nums[i]] > n/2){
                maj = nums[i];
                return maj;
            }
        }
        
        return -1;
    }
};

File: May-LeetCoding-Challange-2020/Week1/169-majority-element.py
class Solution:
    # Sol 1. Hash Map
    def majorityElement(self, nums: List[int]) -> int:
        major_count = len(nums)//2
        c = collections.Counter(nums)
        for i in c: # i is the key of the hash map
            # print(i)
            if c[i] > major_count:
                return i
    # Time Complexity: O(n)
    # Space Complexity: O(n)

    # Sol 2. Brute Force